### ISSUE TITLE
resuming BERT training takes forever...

### ISSUE THREAD

**Issue by tayshinmit** [mosaicml/composer#3649]
We trained Huggingface BERT for approximately 24 hours using 8 A100 GPUs. However, when attempting to resume the training, it appears to use almost all the CPU and GPU resources, but the training never actually begins. We waited for 2 hours with no progress.

Environment below:

Collecting system information...
---------------------------------
System Environment Report        
Created: 2024-10-13 02:49:22 UTC
---------------------------------

PyTorch information
-------------------
PyTorch version: 2.4.0
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: version 3.18.4
Libc version: glibc-2.31

Python version: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.10.0-33-cloud-amd64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 550.90.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
Address sizes:                        46 bits physical, 48 bits virtual
CPU(s):                               96
On-line CPU(s) list:                  0-95
Thread(s) per core:                   2
Core(s) per socket:                   24
Socket(s):                            2
NUMA node(s):                         2
Vendor ID:                            GenuineIntel
CPU family:                           6
Model:                                85
Model name:                           Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:                             7
CPU MHz:                              2200.190
BogoMIPS:                             4400.38
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            1.5 MiB
L1i cache:                            1.5 MiB
L2 cache:                             48 MiB
L3 cache:                             77 MiB
NUMA node0 CPU(s):                    0-23,48-71
NUMA node1 CPU(s):                    24-47,72-95
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat avx512_vnni md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] pytorch-ranger==0.1.1
[pip3] torch==2.4.0
[pip3] torch-optimizer==0.3.0
[pip3] torchaudio==2.4.0
[pip3] torchmetrics==1.4.0.post0
[pip3] torchtext==0.18.0
[pip3] torchvision==0.19.0
[pip3] triton==2.1.0
[conda] blas                      1.0                         mkl    conda-forge
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] libopenvino-pytorch-frontend 2023.2.0             h59595ed_1    conda-forge
[conda] mkl                       2022.1.0           hc2b9512_224  
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch                   2.4.0           py3.10_cuda12.4_cudnn9.1.0_0    pytorch
[conda] pytorch-cuda              12.4                 hc786d27_6    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] pytorch-ranger            0.1.1                    pypi_0    pypi
[conda] torch-optimizer           0.3.0                    pypi_0    pypi
[conda] torchaudio                2.4.0               py310_cu124    pytorch
[conda] torchmetrics              1.4.0.post0              pypi_0    pypi
[conda] torchtext                 0.18.0                   pypi_0    pypi
[conda] torchvision               0.19.0              py310_cu124    pytorch
[conda] triton                    2.1.0                    pypi_0    pypi

Composer information
--------------------
Composer Version: 0.25.0
Composer Commit Hash: None
CPU Model: Intel(R) Xeon(R) CPU @ 2.20GHz
CPU Count: 48
Number of Nodes: 1
GPU Model: NVIDIA A100-SXM4-40GB
GPUs per Node: 1
GPU Count: 1
CUDA Device Count: 8

**Comment by tayshinmit** [comment]
![image](https://github.com/user-attachments/assets/843e3262-30ae-4c81-8cdd-909f043d1540)

Judged by wandbai loging, keep running but not training

![image](https://github.com/user-attachments/assets/f7fe2692-0d16-4981-a8c9-7410837142d8)
For example, loss and other metrics are not logging.

**Comment by tayshinmit** [comment]
Hm..... it is now working. I set load_weights_only=True

**Comment by mvpatel2000** [comment]
What dataloader are you using?

Composer supports deterministic data ordering, which means it has to iterate through the dataloader to resume at the point of last checkpoint. With something like [streaming](https://github.com/mosaicml/streaming), which supports mid-epoch resumption, this happens instantly. For normal dataloaders, you need to iterate.

To confirm this, you can set `spin_dataloaders` ([here](https://github.com/mosaicml/composer/blob/main/composer/trainer/trainer.py#L666)) to `False` and see if that resolves your issue (noting that it will result in sample view duplication).

**Comment by tayshinmit** [comment]
Got it! This is helpful, and I’ve confirmed that the issue is caused by my dataloader trying to find the resume point, which is why it's using CPU resources. Should I switch to streaming, especially since it's resuming mid-epoch? Also, what would the ideal dataloader setup for BERT look like, considering all the data is downloaded locally?

**Comment by snarayan21** [comment]
Hey @tayshinmit, we'd highly recommend streaming for efficient, resumable, and deterministic distributed training. The `StreamingDataset` class is a drop in replacement for a PyTorch iterable-style dataset. Check out the docs [here](https://docs.mosaicml.com/projects/streaming/en/stable/getting_started/quick_start.html).

Essentially how Streaming works is that you will first serialize your dataset to a set of shard files which can be stored remotely. Then, as training progresses, these shard files are progressively downloaded to make samples available for dataloading. The [dataset writing example](https://docs.mosaicml.com/projects/streaming/en/stable/preparing_datasets/basic_dataset_conversion.html) should help you get started!

**Comment by tayshinmit** [comment]
Got it! I have tried streaming with a toy example now, and it works smoothly. Thank you so much—this has been really helpful!
